GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs

```diff
@@ -117,9 +124,81 @@ internal void FillTreeViewNode(TreeNode treeViewNode)
+        internal abstract class NodeBase
+        {
+            /// <summary>The child nodes.</summary>
+            protected internal Nodes Nodes { get; protected set; }
+
+            /// <summary>The corresponding tree node.</summary>
+            protected internal virtual TreeNode TreeViewNode { get; set; }
+
+            /// <summary>
+            /// Marks this node to be included in multi-selection. See <see cref="Select(bool, bool)"/>.
+            /// This is remembered here instead of relying on the status of <see cref="TreeViewNode"/>
+            /// because <see cref="Nodes.FillTreeViewNode(TreeNode)"/> recycles <see cref="TreeNode"/>s
+            /// and may change the association between <see cref="Node"/> and <see cref="TreeNode"/>.
+            /// </summary>
+            protected internal bool IsSelected { get; set; }
+
+            protected internal void Select(bool select, bool includingDescendants = false)
+            {
+                IsSelected = select;
+                ApplyStyle(); // toggle multi-selected node style
+
+                // recursively process descendants if required
+                if (includingDescendants && this.HasChildren())
+                {
+                    foreach (var child in Nodes)
+                    {
+                        child.Select(select, includingDescendants);
+                    }
+                }
+            }
+
+            #region style / appearance
+            protected virtual void ApplyStyle()
+            {
+                SetFont(GetFontStyle());
+                TreeViewNode.ToolTipText = string.Empty;
+            }
+
+            protected virtual FontStyle GetFontStyle()
+                => IsSelected ? FontStyle.Underline : FontStyle.Regular;
+
+            private void SetFont(FontStyle style)
+            {
+                if (style == FontStyle.Regular)
+                {
+                    // For regular, set to null to use the NativeTreeView font
+                    if (TreeViewNode.NodeFont is not null)
+                    {
+                        ResetFont();
+                    }
+                }
+                else
+                {
+                    // If current font doesn't have the input style, get rid of it
+                    if (TreeViewNode.NodeFont is not null && TreeViewNode.NodeFont.Style != style)
+                    {
+                        ResetFont();
+                    }
+
+                    // If non-null, our font is already valid, otherwise create a new one
+                    TreeViewNode.NodeFont ??= new Font(AppSettings.Font, style);
+                }
+            }
+
+            private void ResetFont()
+            {
+                TreeViewNode.NodeFont.Dispose();
+                TreeViewNode.NodeFont = null;
+            }
+            #endregion
+        }
+
+        internal abstract class Tree : NodeBase, IDisposable
```

Comment:
ditto

Reply:
@RussKie Is this about the region? See below. Is this about moving `Tree` into a separate file? I'll do that in a follow-up PR to make reviewing the changes in this on easier.

Reply:
A follow up PR work for me, thank you